### PR TITLE
Analyze all .coffee files when include_paths is ./

### PR DESCRIPTION
**Pull request by BlakeWilliams**
When the `fetch_files` method is passed `./` we run into two problems.
1. It attempts to glob `.//**/*.coffee` which has an extra `/`
2. It prepends `./` to all the results of the glob which aren't present on the results we receive later.

The second issue is a problem because the results returned from coffeelint do not have the `./` prefix which causes none of the files to match leaving us with no results.

We also have to `Dir.chdir` before we do anything otherwise parts of the engine work on the wrong paths.

### CODE REVIEW

**Comment by BlakeWilliams** [comment]
cc @codeclimate/review

File: lib/cc/engine/analyzable_files.rb [hunk 1]
```diff
@@ -19,7 +18,7 @@ def all
       def fetch_files(paths)
         paths.map do |path|
           if path =~ %r{/$}
-            Dir.glob("#{path}/**/*.coffee")
+            Dir.glob("#{path}**/*.coffee")
```
Comment (pbrisbin) [inline lib/cc/engine/analyzable_files.rb:21]: This (and only this) sounds like the bug described in the PR description -- why the other changes? Anything we can break up into separate PRs?

File: lib/cc/engine/coffeelint_results.rb [hunk 2]
```diff
@@ -1,8 +1,7 @@
 module CC
   module Engine
     class CoffeelintResults
-      def initialize(directory, config: nil)
-        @directory = directory
+      def initialize(config: nil)
```
Comment (gdiggs) [inline lib/cc/engine/coffeelint_results.rb]: I know you didn't introduce this, but is there value in this being a key parameter, instead of `config = nil`?

File: lib/cc/engine/analyzable_files.rb (hunk 1 above)
Comment (BlakeWilliams) [inline lib/cc/engine/analyzable_files.rb:21]: I didn't test it, but since we only seemed to `cd` in `CoffeelineResults` I assumed this wouldn't work properly. I can pull out the `cd` changes if that's not the case and we want to extract that into a separate PR.

File: lib/cc/engine/coffeelint_results.rb (hunk 2 above)
Comment (BlakeWilliams) [inline lib/cc/engine/coffeelint_results.rb]: None that I can think of. I'll remove it.

File: lib/cc/engine/analyzable_files.rb (hunk 1 above)
Comment (pbrisbin) [inline lib/cc/engine/analyzable_files.rb:21]: I'm not sure. Just looking at this code in isolation it seems pretty broken: this branch is only taken if `path` ends in a slash, so `#{path}/` is clearly wrong. And that sounds exactly like the `.//` bug you describe.

Comment (BlakeWilliams) [inline lib/cc/engine/analyzable_files.rb:21]: Right, but that also breaks down even without the extra `/` since all the files are then prepended with `./` when they regularly aren't. So when it tries to compare the results from cofeelint and the results of this, this has an extra `./` in front while the others don't.

Comment (BlakeWilliams) [inline lib/cc/engine/analyzable_files.rb:21]: I'll update the description to describe both issues in more detail instead of lumping them into one.

**Comment by ABaldwinHunter** [comment]
@BlakeWilliams if you run `coffeelint <dir>` the tool only analyzes `.coffee` files.

I think it only analyzes non `.coffee` files when passed those directly: `coffeelint <file.rb>`

What if we just checked the include paths and strip any that were `files` && not `.coffee` ?

Analog in radon: https://github.com/codeclimate/radon/blob/master/codeclimate-radon#L21

cc @pbrisbin

**Comment by ABaldwinHunter** [comment]
You would pass in the `include_paths` here:

https://github.com/codeclimate/codeclimate-coffeelint/blob/master/lib/cc/engine/coffeelint_results.rb#L12

e.g.: `coffeelint lib menus other.coffee`

**Comment by BlakeWilliams** [comment]
@ABaldwinHunter I like it, but this engine still supports `exclude_paths` which means I don't think we can do that yet.

Maybe it's worth doing in a follow up PR removing `exclude_paths` and using only `include_paths`?

**Comment by ABaldwinHunter** [comment]
@BlakeWilliams Hm couldn't the `includes` replace `build_files_with_inclusions` or the `.` in the `coffeelint_results` file, if present? How would that impact the exclude paths piece?

**Comment by BlakeWilliams** [comment]
Do you mean pass `AnalyzableFiles#all` to coffeelint?

File: lib/cc/engine/analyzable_files.rb [hunk 3]
```diff
@@ -27,13 +26,17 @@ def fetch_files(paths)
       end
 
       def build_files_with_inclusions(inclusions)
-        fetch_files(inclusions)
+        if inclusions == ["./"]
```
Comment (pbrisbin) [inline lib/cc/engine/analyzable_files.rb:29]: I'm trying to understand why `./` is a special case. What if I `analyze ./foo`, how does that not cause similar problems?

Comment (BlakeWilliams) [inline lib/cc/engine/analyzable_files.rb:29]: That example would also cause similar problems with the current implementation. Thoughts on stripping leading `./` instead?

Comment (pbrisbin) [inline lib/cc/engine/analyzable_files.rb:29]: I guess my suspicion is there's a way to fix this by only changing `#fetch_files`. Do you think that's true?
- `#{path}/` is a clear bug, so let's assume that's fixed
- The remaining bug can be restated as:

`#fetch_files` returns results with `path` pre-pended, but that doesn't match the paths returned by the analysis. Can you unpack in what way they mismatch?

Are the analysis results relative to (each) `path`, or relative to the top-level `code` directory?

Comment (BlakeWilliams) [inline lib/cc/engine/analyzable_files.rb:29]: I think that's correct _most of the time_. It just fails when you get passed `./` (which is default) or the user explicitly passes in anything beginning with `./`.

**Comment by BlakeWilliams** [comment]
Closing in favor of dropping `exclude_paths` completely and supporting only `include_paths`.
